### DISCUSSION TITLE
[accessibility] Listbox: Navigate between options while typing, (like native `<select>` does)

### DISCUSSION THREAD

**Discussion by yaronuliel** [tailwindlabs/headlessui#1957] [Ideas]
In native `<select>` - when the user types on the keyboard while focused on the element, the component looks for the next relevant options by the keystrokes, and replaces the selected value

This feature will help keyboard navigation, which is a huge deal in accessibility. It will also make the user experience better and more expected by resembling the way the browser works.

To resemble browser behavior, keystrokes should:
* Have a debounce before searching for the option (for fast typing)
* Reset the keystrokes after X ms (if the user rests for more than x ms, the next stroke starts a new search)
* When repeating a single keystroke - rotate between options starting with the same letter (<kbd>a</kbd> <kbd>a</kbd> <kbd>a</kbd> sequence will rotate between options starting with `"a"` rather than searching for next option that starts with `"aaa"`)
* Be case insensitive
* Dispatch a `change` event on each change, even during typing (between `keydown` and `keyup` events)
* Skip `disabled` options
* Treat <kbd>Spacebar</kbd> while typing as the space character, and not open the listbox
* Do all the following without opening the listbox
* If the listbox is already open when typing - the behavior is a bit different: The matching option will be highlighted, but not selected (same way navigation with arrows currently works)

Example of what happens in a native select:

https://user-images.githubusercontent.com/5229872/198100075-e9f6376e-9639-4a90-99b8-e1773ace91db.mov

**Comment by vincerubinetti** [comment]
Headless does (at least some of) this when the select dropdown is open, but not when it is collapsed as the native select does. Out of curiosity I looked at [React Aria's select](https://react-spectrum.adobe.com/react-aria/Select.html), and it _does_ do type-ahead when closed, but _not_ when open, oddly. [Material UI](https://mui.com/material-ui/react-select/#basic-select) does it when open. [Vuetify](https://vuetifyjs.com/en/components/selects/) does it when open or closed. Personally I would expect it when open or closed.

**Reply by yaronuliel** [reply]
@vincerubinetti  I think that the library should aim to resemble native behavior, regardless of whether other libraries did it or not

**Reply by vincerubinetti** [reply]
Yes I agree but it is still worthwhile to list what the precedence is for other libraries.

Also, I reported the fact that React Aria doesn't do it when open, and they immediately labeled it as a bug and fixed it already. So, this is hopefully a good motivation for Headless to do the same.
